ast/spec/earthfile.go

```diff
@@ -0,0 +1,77 @@
+package spec
+
+// Earthfile is the AST representation of an Earthfile.
+type Earthfile struct {
+	BaseRecipe     Block           `json:"baseRecipe"`
+	Targets        []Target        `json:"targets"`
+	SourceLocation *SourceLocation `json:"sourceLocation,omitempty"`
+}
+
+// Target is the AST representation of an Earthfile target.
+type Target struct {
+	Name           string          `json:"name"`
+	Recipe         Block           `json:"recipe"`
+	SourceLocation *SourceLocation `json:"sourceLocation,omitempty"`
+}
+
+// Block is the AST representation of a block of statements.
+type Block []Statement
+
+// Statement is the AST representation of an Earthfile statement. Only one field may be
+// filled at one time.
+type Statement struct {
+	Command        *Command        `json:"command,omitempty"`
+	With           *WithStatement  `json:"with,omitempty"`
+	If             *IfStatement    `json:"if,omitempty"`
+	ForIn          *ForInStatement `json:"forIn,omitempty"`
+	SourceLocation *SourceLocation `json:"sourceLocation,omitempty"`
+}
+
+// Command is the AST representation of an Earthfile command.
+type Command struct {
+	Name           string          `json:"name"`
+	Args           []string        `json:"args"`
+	ExecMode       bool            `json:"execMode,omitempty"`
+	SourceLocation *SourceLocation `json:"sourceLocation,omitempty"`
+}
+
+// WithStatement is the AST representation of a with statement.
+type WithStatement struct {
+	Command        Command         `json:"command"`
+	Body           Block           `json:"body"`
+	SourceLocation *SourceLocation `json:"sourceLocation,omitempty"`
+}
+
+// IfStatement is the AST representation of an if statement.
+type IfStatement struct {
+	Expression     []string        `json:"expression"`
+	IfBody         Block           `json:"ifBody"`
+	ElseIf         []ElseIf        `json:"elseIf"`
+	HasElse        bool            `json:"hasElse"`
+	ElseBody       *Block          `json:"elseBody,omitempty"`
+	SourceLocation *SourceLocation `json:"sourceLocation,omitempty"`
+}
+
+// ElseIf is the AST representation of an else if clause.
+type ElseIf struct {
+	Expression     []string        `json:"expression"`
+	Body           Block           `json:"body"`
+	SourceLocation *SourceLocation `json:"sourceLocation,omitempty"`
+}
+
+// ForInStatement is the AST representation of an if statement.
+type ForInStatement struct {
+	Variable       string          `json:"variable"`
+	Expression     []string        `json:"expression"`
+	Body           Block           `json:"body"`
+	SourceLocation *SourceLocation `json:"sourceLocation,omitempty"`
+}
+
+// SourceLocation is an optional reference to the original source code location.
+type SourceLocation struct {
+	File        string `json:"file,omitempty"`
+	StartLine   int    `json:"startLine"`
+	StartColumn int    `json:"startColumn"`
+	EndLine     int    `json:"endLine"`
+	EndColumn   int    `json:"endColumn"`
+}
```

Comment:
Do we want a version here; in case we want to make incompatible changes? Either per-node or whole-AST?

Reply:
Perhaps. I would add it only when we need it, to keep things simple for now. We don't yet store it or use it between different versions of Earthly.